src/container/srv_target.c

```diff
@@ -1035,16 +1046,26 @@ ds_cont_tgt_epoch_aggregate_aggregator(crt_rpc_t *source, crt_rpc_t *result,
 	return 0;
 }
 
-/* iterate all of objects of the container. */
+/* iterate all of objects or uncommitted DTXs of the container. */
 int
-ds_cont_obj_iter(daos_handle_t ph, uuid_t co_uuid,
-		 cont_iter_cb_t callback, void *arg)
+ds_cont_iter(daos_handle_t ph, uuid_t po_uuid, uuid_t co_uuid,
+	     cont_iter_cb_t callback, void *arg, uint32_t ver, uint32_t type)
 {
 	vos_iter_param_t param;
 	daos_handle_t	 iter_h;
 	daos_handle_t	 coh;
 	int		 rc;
 
+	/* Resync DTX before scan others in the container. */
+	if (type != VOS_ITER_DTX) {
+		rc = dtx_resync(ph, po_uuid, co_uuid, ver);
```

Comment:
It is a bit strange to put resync here, I thought the caller should take care of this?

Reply:
The caller is the rebuild logic. You have ever suggested to separate DTX related logic from rebuild logic, then it may be not good to make the rebuild logic to call dtx_resync() directly.

Reply:
This is a general container iteration function.  If the caller needs to make sure no uncommitted data before iteration, then the caller should take care of that, instead of hide this logic inside the iteration.

Sorry, I thought dtx resync is already done before the rebuild? why do we do it again here? Anyway if it is needed, then please do it explicitly during rebuild process.

Reply:
Currently, the rebuild logic needs to guarantee that before the object rebuild, related DTX should be re-synced firstly. Then in my originally implementation, the rebuild_scan_leader() will explicitly call dtx_resync() before rebuild_scanner(), but you said that rebuild logic should not touch that. Then I move it outside rebuild to here. So where is your suggested to trigger the dtx_resync() for rebuild case?

> Sorry, I thought dtx resync is already done before the rebuild? why do we do it again here? 
The DTX re_sync is NOT done yet.

Reply:
oh, I meant recovery process should be done like

all targets finish dtx rebuild,
barrier
then all target start rebuild.

so to split the two phases.  Ideally you may put dtx rebuild phase in dtx module, does not work?

Reply:
Both DTX and object are container based. When rebuild is triggered, there may be multiple containers to be rebuilt. It is unnecessary to make one container's object rebuild to wait for another containers dtx_resync(), right?

Originally, the rebuild_scan_leader explicitly called dtx_resync() before rebuild_scanner(), that did the same as you current suggested: all targets finish dtx rebuild, barrier, then all target start rebuild. But you seems not like such patch and suggest me to change. That is why I move the dtx_resync() at the begging of scan each container.

Anyway, in spite of which solution is used, the caller (either the rebuild logic or the container scan logic) needs to call dtx_resync(). Currently, the dtx_resync() logic is totally inside the DTX module, the dtx_resync() is the API exported to other modules.